paddle/fluid/framework/details/scope_buffered_ssa_graph_executor.cc

```diff
@@ -59,39 +64,53 @@ FeedFetchList ScopeBufferedSSAGraphExecutor::Run(
   }
 }
 
+void ScopeBufferedSSAGraphExecutor::InitVariables() {
+  for (auto &info : tmp_var_infos_) {
+    for (auto &pair : info) {
+      InitializeVariable(pair.first, pair.second);
+    }
+  }
+}
+
 void ScopeBufferedSSAGraphExecutor::DropLocalExeScopes() {
   platform::RecordEvent drop_scope_event("DropLocalExeScopes");
   drop_scope_counter_ = 0;
-  for (auto p : places_) {
+  for (auto &p : places_) {
     platform::DeviceContextPool::Instance().Get(p)->Wait();
   }
 
-  for (auto &scope : local_scopes_) {
-    auto *local_scope_var = scope->FindLocalVar(details::kLocalExecScopeName);
-    if (local_scope_var != nullptr) {
-      auto &local_scope = *local_scope_var->GetMutable<Scope *>();
-      scope->DeleteScope(local_scope);
-      scope->EraseVars({std::string(details::kLocalExecScopeName)});
-      VLOG(3) << "Drop local execution scope: " << local_scope;
+  for (size_t i = 0; i < local_exec_scopes_.size(); ++i) {
+    local_exec_scopes_[i]->ClearWithPreserve(preserve_vars_[i]);
+    for (auto &preserve_var : preserve_vars_[i]) {
+      preserve_var->Clear();
     }
+    VLOG(3) << "Drop local execution scope: " << local_scopes_[i];
   }
 }
 
 void ScopeBufferedSSAGraphExecutor::PrepareLocalExeScopes() {
   // Create local scopes.
+  preserve_vars_.resize(local_scopes_.size());
+  tmp_var_infos_.resize(local_scopes_.size());
+
   for (auto it = local_scopes_.rbegin(); it != local_scopes_.rend(); ++it) {
-    auto &scope = *it;
-    Scope &local_scope = scope->NewScope();
-    *scope->Var(kLocalExecScopeName)->GetMutable<Scope *>() = &local_scope;
+    size_t idx = local_scopes_.size() - 1 - (it - local_scopes_.rbegin());
+    auto *scope = local_scopes_[idx];
+    auto *local_scope = local_exec_scopes_[idx];
 
     for (auto &info : var_infos_) {
-      if (scope->FindVar(info.name_) != nullptr) {
-        continue;
-      }
       if (info.persistable_) {  // Persistable
+        auto var = scope->FindVar(info.name_);
+        if (var != nullptr) {
+          VLOG(2) << info.name_ << " is found in parent scope, skipped";
+          InitializeVariable(var, info.type_);
+          continue;
```

Comment:
1. If `var` is not nullptr, this var may in the `scope` or the `scope`‘s parent.
2. Now that `var` is not nullptr, why do you call `InitializeVariable` here?

Reply:
1. I have changed the log message. Thanks!
2. I have removed `InitializeVariable` when `info.name_` is found in `scope`.

Reply:
1. I have changed the log message. Thanks!
2. I have removed `InitializeVariable` when `info.name_` is found in `scope`.